### ISSUE TITLE
DCP Hash-alternation for ASIC resistance

### ISSUE THREAD

**Issue by mottla** [decred/dcrd#1433]
for personal amusement, I've extended dcrd the following way
one defines from when to when (blockheight) a specific set of cryptographic hashes can be used for mining (for example sha256,sha3,blake2,whirlpool)
if a block is to be mined (or validated) with one of the avaliable hashtypes, I run over the last N blocks, count its occurence and compute its deviation from the expectation value (1/4 th in our example).
if the hashtype one wants to use for the next PoW, is above avg., an exponential function increases the required difficulty (same for validation, e.g consensus can be achieved easily about the difficulty penalty)
so I didnt touch the underlying difficulty mechanism.
I added 1 byte in the block header to indicate, which hashtype has been used for PoW

consequence:
if one has an ASIC for blake2, then he'll quickly will have to do a few thousand times more PoW.
if the last N blocks are blake2 only, then a maximum penalty (depending on a constant-global-variable in the exponent) for additional PoW has to bo payed be the miner.
Cpu/gpu miners would benefit from that and could compete. Adding new hashtypes is extremly easy and would mess up ASICs that support maybe all active hashtipes for a given period.

I thinks this a cool idea and I'd like to discuss the pros and cons of this proposal.
If there is some positive feedback I'll commit the code (yet I'm a bit scared since I've never participated in such a big project before and my go coding style and documentation surely is improveable^^)

**Comment by jrick** [comment]
Decred has been (and with stakeholder approval, will likely remain) an ASIC-friendly project.  In my opinion, aiming for ASIC resistance would needlessly reduce the security of the network, as well as introduce unnecessary complexity to our software.

Rotating PoW algorithms have been attempted by other cryptocurrencies to little success.  In the worst cases, networks have been overtaken by botnets.

Any PoW algorithm change which brings us farther away from the thermodynamic limit is a no vote from me.

**Comment by davecgh** [comment]
A quick minor note that for Decred, they are [DCPs - Decred Change Proposal](https://github.com/decred/dcps).

As far as the proposal itself, I'm in agreement with @jrick on this topic.  Not only are rotating PoW algorithms not particularly effective, ASIC resistance, in my opinion, is not at all a desirable property for Decred for several reasons anyway.  Rather than repeat them here, I would like to refer to [my responses to the a reddit thread regarding this topic](https://www.reddit.com/r/decred/comments/7dedss/asics_or) which provides a quite thorough treatment.

That said, since Decred is a stakeholder-driven project, the way that a change such as this would have to proceed is that, per my initial link to the DCP process is:

```
It is important to note that DCPs are intended to be the end result of an off-chain proposal and voting system that deals with more generalized proposal submissions. It is through that system that all initial proposals for consensus or protocol changes will be brought to life, undergo collaboration, and be subject to an initial community vote to determine if the work necessary to create a working implementation, its associated DCP, and an on-chain vote for the rule change should be performed. 
```

Politeia (Pi) is the system that text is referring to and is currently undergoing testing prior to launch.

So, in other words, once Pi is launched, you would be able to submit this as a proposal to see if the stakeholders would be interested in proceeding to the stage where the code and a full-blown DCP would be created to have a final on-chain vote of the actual implementation.

**Comment by davecgh** [comment]
Closing this now the Politeia has launched.  As mentioned, any changes such as this would need a full-blown proposals through Pi so it could be voted on by the stakeholders.
